### ISSUE TITLE
Hybrid Feature Bugs [local]

### ISSUE THREAD

**Issue by maysunfaisal** [eclipse/codewind#483]
- [ ] Deleting a file from my local project in the codewind-workspace does not get reflected in the PFE container workspace
- [ ] Getting capabilities and logs initially fails for all project types, which has been seen in the past when doing requests to Portal before Turbine has "created" the project 
```
swifty: Getting capabilities failed: Project 76178900-da55-11e9-990c-096baec0f02f not found, internal error occurred. Filewatcher status code 404
swifty: checking available logs failed: Project 76178900-da55-11e9-990c-096baec0f02f not found, internal error occurred. Filewatcher status code 404
```
- [ ] Go and Python projects are not given their correct language: 
![image](https://user-images.githubusercontent.com/29612372/65181117-55897a00-da2c-11e9-979a-3915f411a055.png)
- [ ] Some upload requests get rejected with 413 errors [comment](https://github.com/eclipse/codewind/issues/483#issuecomment-532859497)
- [ ] Looking at the code in https://github.com/eclipse/codewind-vscode/blob/e41afb3a0f63a85036ec44d082be33365194f639/dev/src/codewind/project/Requester.ts#L184 it looks like individual files are being uploaded. This could be a performance problem when dealing with projects that have many files.

**Comment by maysunfaisal** [comment]
Docker projects seem to be fine - Python, Go

**Comment by tetchel** [comment]
In the midst of the remote uploads, I see 413 errors:

```
[INFO: 16:40:31.893 Requester.js:192]: requestUploadsRecursively for 8e8c7af0-da54-11e9-990c-096baec0f02f at /Users/tim/programs/codewind/codewind-workspace/microproj/mc-target/liberty/wlp/usr/servers/defaultServer/apps/expanded/javametrics-dash.war/graphmetrics/bootstrap-3.3.7-dist/css
[INFO: 16:40:31.906 Requester.js:230]: Remote upload response OK
[DBUG: 16:40:31.933 Requester.js:203]: StatusCodeError: 413 - "request entity too large"
	at new StatusCodeError (/Users/tim/programs/codewind-vscode/dev/node_modules/request-promise-core/lib/errors.js:32:15)
	at Request.module.exports.plumbing.callback (/Users/tim/programs/codewind-vscode/dev/node_modules/request-promise-core/lib/plumbing.js:104:33)
	at Request.RP$callback [as _callback] (/Users/tim/programs/codewind-vscode/dev/node_modules/request-promise-core/lib/plumbing.js:46:31)
	at Request.init.self.callback (/Users/tim/programs/codewind-vscode/dev/node_modules/request/request.js:185:22)
	at Request.emit (events.js:182:13)
	at Request.<anonymous> (/Users/tim/programs/codewind-vscode/dev/node_modules/request/request.js:1161:10)
	at Request.emit (events.js:182:13)
	at IncomingMessage.<anonymous> (/Users/tim/programs/codewind-vscode/dev/node_modules/request/request.js:1083:12)
	at Object.onceWrapper (events.js:273:13)
	at IncomingMessage.emit (events.js:187:15)
	at endReadableNT (_stream_readable.js:1092:12)
	at process._tickCallback (internal/process/next_tick.js:63:19)
[INFO: 16:40:31.934 Requester.js:192]: requestUploadsRecursively for 8e8c7af0-da54-11e9-990c-096baec0f02f at /Users/tim/programs/codewind/codewind-workspace/microproj/mc-target/liberty/wlp/usr/servers/defaultServer/apps/expanded/javametrics-dash.war/graphmetrics/bootstrap-3.3.7-dist/js
[INFO: 16:40:31.941 Requester.js:230]: Remote upload response OK
```

**Comment by hhellyer** [comment]
@maysun  
> Deleting a file from my local project in the codewind-workspace does not get reflected in the PFE container workspace
This should be fixed by using the fixed branch but I am about to add some improved syncing so we only send changed files and only delete files that no longer exist locally.

As you say, I think it's because Turbine doesn't know about the project yet. I'll check to see if the project state indicates it's not ready yet and then we can check that in the plugins before we query the capabilities and logs. 

I'm not sure there's anything that should affect the project types, is that working in non-remote mode?

I'll take a look at the 413 error, I tried creating a liberty project in my workspace and didn't see it which is annoying. I'll look into if we need to break up larger files and if there's a way to query the http server to find out how much data it can handle in one request so we know how big the chunks should be.

Individual files are being uploaded. That does mean a big upload on project creation but generally users will only change one file in their editor at a time. (Things like multi-file find/replace might change a lot of files at once.)
I don't want to optimise that unless we really see it being a problem things like only sending diffs are tricky and fragile. I'm also not convinced there'll be much benefit - source files are usually small enough that we shouldn't worry, big binary files like jars don't diff.
The first pass, which I'll raise a PR for shortly, is to only upload the files that have changed since the last upload.

**Comment by maysunfaisal** [comment]
I am also seeing this error on a Spring manual build:

```
[19/09/19 16:02:14 /portal/middleware/errorHandler.js] [ERROR] { error:
   { name: 'PayloadTooLargeError',
     message: 'request entity too large' } }
[19/09/19 16:02:15 /portal/middleware/errorHandler.js] [ERROR] { error:
   { name: 'PayloadTooLargeError',
     message: 'request entity too large' } }
[19/09/19 16:02:16 /portal/routes/projects/remoteBind.route.js] [ERROR] { Error: Command failed: rm -rf /codewind-workspace/spr123/* /codewind-workspace/spr123/.[!.]*
rm: cannot remove '/codewind-workspace/spr123/.m2/repository/org/apache/maven': Directory not empty

    at ChildProcess.exithandler (child_process.js:294:12)
    at ChildProcess.emit (events.js:198:13)
    at maybeClose (internal/child_process.js:982:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:259:5)
  killed: false,
  code: 1,
  signal: null,
  cmd:
   'rm -rf /codewind-workspace/spr123/* /codewind-workspace/spr123/.[!.]*',
  stdout: '',
  stderr:
   'rm: cannot remove \'/codewind-workspace/spr123/.m2/repository/org/apache/maven\': Directory not empty\n' }
```

![Screen Shot 2019-09-19 at 12 02 37 PM](https://user-images.githubusercontent.com/31771087/65261175-a5704b80-dad5-11e9-8eff-86df8685d403.png)

**Comment by rajivnathan** [comment]
@hhellyer 
>Individual files are being uploaded. That does mean a big upload on project creation but generally users will only change one file in their editor at a time. (Things like multi-file find/replace might change a lot of files at once.)

Agree that uploading individual files is okay for update scenarios, but even then we'd probably want to upload a set of changes as a batch so that a build won't begin with a partial change set.

I'm also concerned about the performance impact on project creation in a local -> kube case where latency can be poor. We could look at zipping up the project folder and transfer the single file just for the first time instead of transferring each file individually.

**Comment by hhellyer** [comment]
I've made an update to pfe so that in remote mode we only need to send changed files. This is in `eclipse/codewind#master`. This requires updated versions of the hybrid branches for `codewind-vscode` and `codewind-eclipse` which I've also pushed to the `hybrid` branch on both those repositories.

@rajivnathan - I also added log messages that include the time taken to do the uploads so we can evaluate the performance.

**Comment by johnmcollier** [comment]
@hhellyer 
Not necessarily bugs, but a preliminary list of things that may need to be done or checked for Che on hybrid:
```
- Remove Che workspace volume mount from PFE??
  - Sounds like it's not needed any more and won't work with the current hybrid architecture
- Create separate, RWX volume for PFE (and the projects that require it, i.e Liberty & spring)
  - Liberty and Spring projects would need to mount this volume instead
- API needed to ensure the docker registry and repository list are set in PFE
- Concerns about network latency and NFS
  - Not necessarily a todo, but need to see how well the hybrid changes handle when NFS is used
```

**Comment by tobespc** [comment]
closing as issues dealt with

**Comment by Abhik1998** [comment]
![image](https://user-images.githubusercontent.com/31558638/68250715-8149d900-0047-11ea-8864-2738db314cfc.png)
I am getting this error

**Comment by rtaniwa** [comment]
@Abhik1998 should this last comment be an issue? Isn't clear to me if you added the comment here because you thought it was related to this issue or not?

**Comment by Abhik1998** [comment]
I thought It was related to this issue.


> * [ ]  Getting capabilities and logs initially fails for all project types, which has been seen in the past when doing requests to Portal before Turbine has "created" the project
> 
> ```
> swifty: Getting capabilities failed: Project 76178900-da55-11e9-990c-096baec0f02f not found, internal error occurred. Filewatcher status code 404
> swifty: checking available logs failed: Project 76178900-da55-11e9-990c-096baec0f02f not found, internal error occurred. Filewatcher status code 404
> ```

**Comment by tetchel** [comment]
Thanks for commenting - It looks like you hit this issue on our production 0.5 release. This behaviour has been reworked in our next version and should not be seen in 0.6. If you still see it then, please open a new issue.
